x/ccv/provider/ibc_middleware.go

```diff
@@ -0,0 +1,239 @@
+func (im IBCMiddleware) OnChanCloseConfirm(
+	ctx sdk.Context,
+	portID,
+	channelID string,
+) error {
+	return im.app.OnChanCloseConfirm(ctx, portID, channelID)
+}
+
+// OnRecvPacket executes the IBC transfer. In case of success,
+// it verifies if the packet sender is a consumer chain
+// and if the received IBC coin is whitelisted. In such instances,
+// it appends the coin to the consumer's chain allocation record
+func (im IBCMiddleware) OnRecvPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	relayer sdk.AccAddress,
+) exported.Acknowledgement {
+	// executes the IBC transfer OnRecv logic
+	ack := im.app.OnRecvPacket(ctx, packet, relayer)
+	if ack.Success() {
+
+		// check if the packet sender is a consumer chain
+		chainID, err := im.keeper.IdentifyConsumerChainIDFromIBCPacket(ctx, packet)
+		if err != nil {
+			return ack
+		}
+
+		// At this point the transfer was successful,
+		// so the IBC packet can be safely deserialized
+		var data ibctransfertypes.FungibleTokenPacketData
+		_ = types.ModuleCdc.UnmarshalJSON(packet.GetData(), &data)
+
+		// check if the recipient is the consumer reward's pool address
+		receiver, _ := sdk.AccAddressFromBech32(data.Receiver)
+		if receiver.String() != im.keeper.GetConsumerRewardsPoolAddressStr(ctx) {
+			return ack
+		}
+
+		coinDenom := getProviderDenom(data.Denom, packet)
+		coinAmt, _ := math.NewIntFromString(data.Amount)
+
+		// Iterate over the whitelisted consumer denoms
+		for _, denom := range im.keeper.GetAllConsumerRewardDenoms(ctx) {
+			// if the coin's denom is found,
+			// update the consumer chain rewards allocation
+			if denom == coinDenom {
+				alloc := im.keeper.GetConsumerRewardsAllocation(ctx, chainID)
+				alloc.Rewards = alloc.Rewards.Add(
+					sdk.NewDecCoinsFromCoins(sdk.Coin{
+						Denom:  coinDenom,
+						Amount: coinAmt,
+					})...)
+				im.keeper.SetConsumerRewardsAllocation(ctx, chainID, alloc)
+
+				break
+			}
+		}
+	}
+
+	return ack
+}
+
+// OnAcknowledgementPacket implements the IBCMiddleware interface
+// If fees are not enabled, this callback will default to the ibc-core packet callback
+func (im IBCMiddleware) OnAcknowledgementPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	acknowledgement []byte,
+	relayer sdk.AccAddress,
+) error {
+	return im.app.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer)
+}
+
+// OnTimeoutPacket implements the IBCMiddleware interface
+// If fees are not enabled, this callback will default to the ibc-core packet callback
+func (im IBCMiddleware) OnTimeoutPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	relayer sdk.AccAddress,
+) error {
+	// call underlying callback
+	return im.app.OnTimeoutPacket(ctx, packet, relayer)
+}
+
+// SendPacket implements the ICS4 Wrapper interface
+func (im IBCMiddleware) SendPacket(
+	sdk.Context,
+	*capabilitytypes.Capability,
+	string,
+	string,
+	clienttypes.Height,
+	uint64,
+	[]byte,
+) (uint64, error) {
+	return 0, nil
```

Comment:
Why not have `return im.SendPacket(...)` here instead of `return 0, nil`?

Reply:
`return im.SendPacket(...) ` creates an infinite recursive call...
AFAIU in IBC transfer app module doesn't implement `SendPacket` and sends packet using [ics4wrapper](https://github.com/cosmos/ibc-go/blob/v7.3.2/modules/apps/transfer/keeper/relay.go#L133) instead.